airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -222,224 +212,356 @@ def _iter_upstream_conditions() -> Iterator[ColumnOperators]:
+                            f"upstream_states={upstream_states}, "
+                            f"upstream_task_ids={task.upstream_task_ids}"
+                        )
                     )
-                )
-            elif upstream_setup and not success_setup >= 1:
-                yield self._failing_status(
-                    reason=(
-                        f"Task's trigger rule '{trigger_rule}' requires at least one upstream setup task be "
-                        f"successful, but found {upstream_setup - success_setup} task(s) that were not. "
-                        f"upstream_states={upstream_states}, "
-                        f"upstream_task_ids={task.upstream_task_ids}"
+                elif upstream_setup is None:  # for now, None only happens in mapped case
+                    yield self._failing_status(
+                        reason=(
+                            f"Task's trigger rule '{trigger_rule}' cannot have mapped tasks as upstream. "
+                            f"upstream_states={upstream_states}, "
+                            f"upstream_task_ids={task.upstream_task_ids}"
+                        )
+                    )
+                elif upstream_setup and not success_setup >= 1:
+                    yield self._failing_status(
+                        reason=(
+                            f"Task's trigger rule '{trigger_rule}' requires at least one upstream setup task "
+                            f"be successful, but found {upstream_setup - success_setup} task(s) that were "
+                            f"not. upstream_states={upstream_states}, "
+                            f"upstream_task_ids={task.upstream_task_ids}"
+                        )
                     )
-                )
-        else:
-            yield self._failing_status(reason=f"No strategy to evaluate trigger rule '{trigger_rule}'.")
+            else:
+                yield self._failing_status(reason=f"No strategy to evaluate trigger rule '{trigger_rule}'.")
+
+        # this dictionary enables use of task_id as the key for lru caching
+        # in function _get_relevant_upstream_map_indexes
+        tasks_dict = {}
+
+        if not ti.task.is_teardown:
+            # a teardown cannot have any indirect setups
+            relevant_setups = {t.task_id: t for t in ti.task.get_upstreams_only_setups()}
+            if relevant_setups:
+                # set tasks_dict to contain relevant setups
+                # _get_relevant_upstream_map_indexes will use this to look up the task_id
+                tasks_dict = relevant_setups
+                for status, changed in _evaluate_setup_constraint(relevant_setups=relevant_setups):
+                    yield status
+                    if not status.passed and changed:
+                        # no need to evaluate trigger rule; we've already marked as skipped or failed
+                        return
+
+        task = ti.task
+        trigger_rule = task.trigger_rule
```

Comment:
I feel these two should be put back to near the top of the function, or at least before the (nested) function that uses them.

Reply:
yeah true